cmd/gardenlet/app/migration.go

```diff
@@ -237,3 +245,161 @@ func cleanupNginxConfigMaps(ctx context.Context, client client.Client) error {
+			return nil, nil, err
+		}
+
+		return managedResource, client.MergeFrom(managedResource.DeepCopy()), nil
+	}
+
+	unignoreManagedResource := func(ctx context.Context, log logr.Logger, cluster string) {
+		managedResource, managedResourcePatch, err := getManagedResourceWithPatch(ctx, cluster)
+		if err != nil {
+			log.Error(err, "Failed to get ManagedResource for Prometheus to unignore", "cluster", cluster)
+			return
+		}
+
+		if value, ok := managedResource.Annotations[resourcesv1alpha1.Ignore]; ok && value == "true" {
+			delete(managedResource.Annotations, resourcesv1alpha1.Ignore)
+			if err := seedClient.Patch(ctx, managedResource, managedResourcePatch); err != nil {
+				log.Error(err, "Failed to unignore ManagedResource for Prometheus", "cluster", cluster)
+			}
+		}
+	}
+
+	getPrometheusWithPatch := func(ctx context.Context, namespace string) (*monitoringv1.Prometheus, client.Patch, error) {
+		prometheus := &monitoringv1.Prometheus{}
+		if err := seedClient.Get(ctx, client.ObjectKey{Namespace: namespace, Name: "shoot"}, prometheus); err != nil {
+			return nil, nil, err
+		}
+
+		return prometheus, client.MergeFrom(prometheus.DeepCopy()), nil
+	}
+
+	needsInitContainer := func(prometheus *monitoringv1.Prometheus) bool {
+		for _, initContainer := range prometheus.Spec.InitContainers {
+			if initContainer.Name == "cleanup-obsolete-folder" {
+				return false
+			}
+		}
+
+		return true
+	}
+
+	waitUntilCleanedUp := func(ctx context.Context, namespace string) error {
+		return retry.UntilTimeout(ctx, 10*time.Second, 10*time.Minute, func(ctx context.Context) (done bool, err error) {
+			pod := &corev1.Pod{ObjectMeta: metav1.ObjectMeta{Namespace: namespace, Name: "prometheus-shoot-0"}}
+			if err := seedClient.Get(ctx, client.ObjectKeyFromObject(pod), pod); err != nil {
+				return retry.MinorError(err)
+			}
+
+			var hasInitContainer bool
+			for _, initContainer := range pod.Spec.InitContainers {
+				if initContainer.Name == "cleanup-obsolete-folder" {
+					hasInitContainer = true
+					break
+				}
+			}
+
+			if !hasInitContainer {
+				return retry.MinorError(fmt.Errorf("prometheus-shoot-0 pod does not have the cleanup-obsolete-folder init container"))
+			}
+
+			if err := health.CheckPod(pod); err != nil {
+				return retry.MinorError(fmt.Errorf("prometheus-shoot-0 pod is not healthy: %w", err))
+			}
+
+			return retry.Ok()
+		})
+	}
+
+	log.Info("Clean up Prometheus obsolete folders")
+
+	clusterList := &extensionsv1alpha1.ClusterList{}
+	if err := seedClient.List(ctx, clusterList); err != nil {
+		log.Error(err, "Failed to list clusters while cleaning up Prometheus obsolete folders")
+		return
+	}
+
+	for _, cluster := range clusterList.Items {
+		unignoreManagedResource(ctx, log, cluster.Name)
+
+		tasks = append(tasks, func(ctx context.Context) error {
+			managedResource, managedResourcePatch, err := getManagedResourceWithPatch(ctx, cluster.Name)
```

Comment:
What if the managed resource is not found/does not exist? This should be tolerated, I guess